<a href="https://colab.research.google.com/github/joanasr15/python_data/blob/main/Sorting_and_cleaning.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Sorting and cleaning
---



In order to effectively analyse a dataset, often we need to prepare it first.
Before a dataset is ready to be analysed we might need to:  

* sort the data (can be a series or dataframe)  
* remove any NaN values or drop NA values   
* remove duplicate records (identical rows)  
* normalise data in dataframe columns so that has a common scale [reference](https://towardsai.net/p/data-science/how-when-and-why-should-you-normalize-standardize-rescale-your-data-3f083def38ff#:~:text=Similarly%2C%20the%20goal%20of%20normalization,dataset%20does%20not%20require%20normalization.&text=So%20we%20normalize%20the%20data,variables%20to%20the%20same%20range.)

## Sorting the data  
---


Typically we want to sort data by the values in one or more columns in the dataframe  

To sort the dataframe by series we use the pandas function **sort_values()**.  

By default `sort_values()` sorts into ascending order.

* sort by a single column e.g.
  * `df.sort_values("Make") `
* sort by multiple columns e.g.
  * `df.sort_values(by = ["Model", "Make"]) `
    * this sorts by Model, then my Make
* sort in *descending* order
  * `df.sort_values(by = "Make", ascending = False)`
  * `df.sort_values(by = ["Make", "Model"], ascending = False])`  

Dataframes are mostly immutable, changes like sort_values do not change the dataframe permanently, they just change it for the time that the instruction is being used.

`df.sort_values(by='Make')` *dataframe is now in sorted order and can be copied to a new dataframe*  
`df` *original dataframe, df, will be as it was - unsorted*

To split the dataframe after sorting, do this in the same instruction, e.g.:

`df.sort_values(by = ["Make", "Model"], ascending = False])[["Make", "Model"]]`

This sorts on Make and then Model in descending order, then splits off the Make and Model columns.

`df.sort_values(by = ["Make", "Model"], ascending = False])[["Make", "Model"]].head()`

This sorts on Make and then Model, then splits off the Make and Model columns and then splits off the first 5 rows.

### Exercise 1 - get data, sort by happiness score
---

Read data from the Excel file on Happiness Data at this link: https://github.com/futureCodersSE/working-with-data/blob/main/Happiness-Data/2015.xlsx?raw=true

Display first 5 rows of data  

The data is currently sorted by Happiness Rank...
*  sort the data by Happiness Score in ascending order
*  display sorted table

**Test output**:  
The lowest score (displayed first) is 2.839, Togo  
The highest score (displayed last) is 7.587, Switzerland  



In [1]:
import pandas as pd

url =  "https://github.com/futureCodersSE/working-with-data/blob/main/Happiness-Data/2015.xlsx?raw=true"

df = pd.read_excel(url, sheet_name="Happiness-Data-2015")

#sort the data by Happiness Score in ascending order
#display sorted table

display(df.head(5))

display(df.sort_values("Happiness Score"))

Country          Region  Happiness Rank  Happiness Score  \
0  Switzerland  Western Europe               1            7.587   
1      Iceland  Western Europe               2            7.561   
2      Denmark  Western Europe               3            7.527   
3       Norway  Western Europe               4            7.522   
4       Canada   North America               5            7.427   

   Standard Error  Economy (GDP per Capita)   Family  \
0         0.03411                   1.39651  1.34951   
1         0.04884                   1.30232  1.40223   
2         0.03328                   1.32548  1.36058   
3         0.03880                   1.45900  1.33095   
4         0.03553                   1.32629  1.32261   

   Health (Life Expectancy)  Freedom  Trust (Government Corruption)  \
0                   0.94143  0.66557                        0.41978   
1                   0.94784  0.62877                        0.14145   
2                   0.87464  0.64938                        0.48357   
3                   0.88521  0.66973                        0.36503   
4                   0.90563  0.63297                        0.32957   

   Generosity  Dystopia Residual  
0     0.29678            2.51738  
1     0.43630            2.70201  
2     0.34139            2.49204  
3     0.34699            2.46531  
4     0.45811            2.45176

Country                           Region  Happiness Rank  \
157         Togo               Sub-Saharan Africa             158   
156      Burundi               Sub-Saharan Africa             157   
155        Syria  Middle East and Northern Africa             156   
154        Benin               Sub-Saharan Africa             155   
153       Rwanda               Sub-Saharan Africa             154   
..           ...                              ...             ...   
4         Canada                    North America               5   
3         Norway                   Western Europe               4   
2        Denmark                   Western Europe               3   
1        Iceland                   Western Europe               2   
0    Switzerland                   Western Europe               1   

     Happiness Score  Standard Error  Economy (GDP per Capita)   Family  \
157            2.839         0.06727                   0.20868  0.13995   
156            2.905         0.08658                   0.01530  0.41587   
155            3.006         0.05015                   0.66320  0.47489   
154            3.340         0.03656                   0.28665  0.35386   
153            3.465         0.03464                   0.22208  0.77370   
..               ...             ...                       ...      ...   
4              7.427         0.03553                   1.32629  1.32261   
3              7.522         0.03880                   1.45900  1.33095   
2              7.527         0.03328                   1.32548  1.36058   
1              7.561         0.04884                   1.30232  1.40223   
0              7.587         0.03411                   1.39651  1.34951   

     Health (Life Expectancy)  Freedom  Trust (Government Corruption)  \
157                   0.28443  0.36453                        0.10731   
156                   0.22396  0.11850                        0.10062   
155                   0.72193  0.15684                        0.18906   
154                   0.31910  0.48450                        0.08010   
153                   0.42864  0.59201                        0.55191   
..                        ...      ...                            ...   
4                     0.90563  0.63297                        0.32957   
3                     0.88521  0.66973                        0.36503   
2                     0.87464  0.64938                        0.48357   
1                     0.94784  0.62877                        0.14145   
0                     0.94143  0.66557                        0.41978   

     Generosity  Dystopia Residual  
157     0.16681            1.56726  
156     0.19727            1.83302  
155     0.47179            0.32858  
154     0.18260            1.63328  
153     0.22628            0.67042  
..          ...                ...  
4       0.45811            2.45176  
3       0.34699            2.46531  
2       0.34139            2.49204  
1       0.43630            2.70201  
0       0.29678            2.51738  

[158 rows x 12 columns]

### Exercise 2 - sort by multiple columns, display the first 5 rows
---

1. sort the data by Economy (GDP per Capita) and Health (Life Expectancy) in ascending order
2. display the first 5 rows of sorted data

**Test output**:  
Records 122, 127, 147, 100, 96

In [2]:
import pandas as pd

url =  "https://github.com/futureCodersSE/working-with-data/blob/main/Happiness-Data/2015.xlsx?raw=true"

df = pd.read_excel(url, sheet_name="Happiness-Data-2015")

#sort the data by Economy (GDP per Capita) and Health (Life Expectancy) in ascending order
#display the first 5 rows of sorted data

display(df.sort_values(by = ["Health (Life Expectancy)","Economy (GDP per Capita)"]).head(5))

Country              Region  Happiness Rank  \
122              Sierra Leone  Sub-Saharan Africa             123   
127                  Botswana  Sub-Saharan Africa             128   
147  Central African Republic  Sub-Saharan Africa             148   
100                 Swaziland  Sub-Saharan Africa             101   
96                    Lesotho  Sub-Saharan Africa              97   

     Happiness Score  Standard Error  Economy (GDP per Capita)   Family  \
122            4.507         0.07068                   0.33024  0.95571   
127            4.332         0.04934                   0.99355  1.10464   
147            3.678         0.06112                   0.07850  0.00000   
100            4.867         0.08742                   0.71206  1.07284   
96             4.898         0.09438                   0.37545  1.04103   

     Health (Life Expectancy)  Freedom  Trust (Government Corruption)  \
122                   0.00000  0.40840                        0.08786   
127                   0.04776  0.49495                        0.12474   
147                   0.06699  0.48879                        0.08289   
100                   0.07566  0.30658                        0.03060   
96                    0.07612  0.31767                        0.12504   

     Generosity  Dystopia Residual  
122     0.21488            2.51009  
127     0.10461            1.46181  
147     0.23835            2.72230  
100     0.18259            2.48676  
96      0.16388            2.79832

### Exercise 3 - sorting in descending order
---

Sort the data by Freedom and Trust (Government Corruption) in descending order and show the Country and Region only for the last five rows

**Test output**:
136, 117, 95, 101, 111 Country and Region columns


In [3]:
import pandas as pd

url =  "https://github.com/futureCodersSE/working-with-data/blob/main/Happiness-Data/2015.xlsx?raw=true"

df = pd.read_excel(url, sheet_name="Happiness-Data-2015")

#Sort the data by Freedom and Trust (Government Corruption) in descending order and show the Country and Region only for the last five rows

display(df.sort_values(by = ["Freedom","Trust (Government Corruption)"], ascending = False).tail(5))

Country                           Region  Happiness Rank  \
136                  Angola               Sub-Saharan Africa             137   
117                   Sudan               Sub-Saharan Africa             118   
95   Bosnia and Herzegovina       Central and Eastern Europe              96   
101                  Greece                   Western Europe             102   
111                    Iraq  Middle East and Northern Africa             112   

     Happiness Score  Standard Error  Economy (GDP per Capita)   Family  \
136            4.033         0.04758                   0.75778  0.86040   
117            4.550         0.06740                   0.52107  1.01404   
95             4.949         0.06913                   0.83223  0.91916   
101            4.857         0.05062                   1.15406  0.92933   
111            4.677         0.05232                   0.98549  0.81889   

     Health (Life Expectancy)  Freedom  Trust (Government Corruption)  \
136                   0.16683  0.10384                        0.07122   
117                   0.36878  0.10081                        0.14660   
95                    0.79081  0.09245                        0.00227   
101                   0.88213  0.07699                        0.01397   
111                   0.60237  0.00000                        0.13788   

     Generosity  Dystopia Residual  
136     0.12344            1.94939  
117     0.19062            2.20857  
95      0.24808            2.06367  
101     0.00000            1.80101  
111     0.17922            1.95335

# Cleaning the data

Data comes from a range of sources:  forms, monitoring devices, etc.  There will often be missing values, duplicate records and values that are incorrectly formatted.  These can affect summary statistics and graphs plotted from the data.

Techniques for data cleansing include:
*  removing records with missing or null data (NaN, NA, "")
*  removing duplicate rows (keeping just one, either the first or the last)

Removal of rows according to criteria, or of columns are other ways that data might be cleaned up.  


---

## Removing NaN/Dropping NA values

pandas have functions for checking a dataframe, or column, for null values, checking a column for missing values, and functions for dropping all rows that contain null values.

* check for NA/NaN/missing values across dataframe (returns True if NA values exist)  
  `df.isnull().values.any()`  

* check for NA/NaN/missing values in specific column  
  `df["Make"].isnull().values.any()`  

* drop all rows that have NA/NaN values   
  `df.dropna()`  

* drop rows where NA/NaN values exist in specific columns  
  `df.dropna(subset = ["Make", "Model"])`  

### Exercise 4 - check for null values
---

1. read data from the file housing_in_london_yearly_variables.csv from this link: https://raw.githubusercontent.com/futureCodersSE/working-with-data/main/Data%20sets/housing_in_london_yearly_variables.csv
2. check if any NA values exist in the dataframe and print the result
3. use df.info() to see which columns have null entries (*Hint: if the non-null count is less than total entries, column contains missing/NA entries*)  

**Test output**:
True
.info shows median_salary, life_satisfaction, recycling_pct, population_size, number_of_jobs, area_size, no_of_houses all less than total rows (1071)



### Exercise 5 - remove null values
---

1. remove rows with NA values for `life_satisfaction` (use [ ] even if only one column in list)
2. remove all NA values across whole dataframe

**Test output**:  
1.  Row count reduced to 352 rows
2.  Row count reduced to 267 rows

## Dropping duplicates
---

* To remove duplicate rows based on duplication of values in all columns  
  `df.drop_duplicates()`  

* To remove rows that have duplicate entries in a specified column  
  `df.drop_duplicates(subset = ['Make'])`  

* To remove rows that have duplicate entries in multiple columns  
  `df.drop_duplicates(subset = ['Make', 'Model'])`

* Remove duplicate rows keeping the last instance rather than the first (default):  
  `df.drop_duplicates(keep='last')`  

### Exercise 6 - Removing duplicate entries
---

remove duplicate `area` entries keeping first instance  

**Test output**:  
 Dataframe now contains 50 rows all with date 1999-12-*01*

# Normalising Data  
When we normalise data, we remodel a numeric column in a dataframe to be on a standard scale (e.g. 0 or 1).   

For example if we had a column of BMI scores, we could normalise that column so that all scores greater than or equal to 25 were recoded to the value 1 (bad) and all scores less than 25 were recoded to 0 (good).  

To normalise we need to:
*   write a function, with the dataframe as a parameter, which will look at each row in dataframe column and return either a value in the normalised scale (e.g. 0,1 or 1,2,3,4) depending on that value.

For example:  
```
def normalise_bmi(df):
  if df['bmi'] >= 25:
    return 1
  else:
    return 0

df["bmi"] = df.apply(normalise_bmi, axis=1)
```
This code reassigns the values in the column "bmi" by sending each row one after the other to the normalise_bmi function, which will check the value in the "bmi" column and return either 0 or 1 depending on the value in the "bmi" column.

### Exercise 7 - normalise data set
---

Create a function called **normalise_income(df)** that will return the values 1, 2 or 3 to represent low income, middle income and high income.  If the value in `df['median_salary']` is less than 27441 (the median), return 1, otherwise if it is less than 30932 (the upper quartile) return 2 and otherwise return 3.

Apply the normalise_income(df) function to the `median_salary` column.

*NOTE:  this operation will change the original dataframe so if you run it twice, everything in the median_salary column will change to 1 (as it had already been reduced to 1, 2 or 3 - if this happens, run the code in Exercise 4 again to get the original data again from the file.*

**Test output**:  
The maximum value of the column df['median_salary'] will be 3 and the minimum value will be 1  

### Exercise 8 - normalise the number of jobs column
---

Using what you have learnt from Exercise 7:  
*  use `df.describe()` to find the median, upper quartile and maximum for the number_of_jobs column  
*  create a function called **normalise_jobs(df)** that will return 1 if the `number_of_jobs` is below the median, 2 if the `number_of_jobs` is below the upper quartile or 3 otherwise.
*  normalise the `number_of_jobs` column by applying the function `normalise_jobs`.

**Test output**:  
The maximum value of the column df['number_of_jobs'] will be 3 and the minimum value will be 1  

## Exercise 9 - normalise into a new column
---

Create a new function and code to normalise the `no_of_houses` column BUT this time, instead of assigning the result to `df['no_of_houses']` assign it to a new column called `df['housing_volume']`

**Test output**:  
The maximum value of the column df['housing_volume'] will be 3 and the minimum value will be 1

### Exercise 10 - normalise boroughs
---

Normalise the `area_size` column so that all values below mean are represented as 0 and otherwise are 1.  Assign the output to a new column called `area_size_normalised`.  

**Test output**:  
`area_size_normalised` column will contain both 0s and 1s.  The position of the first row with value 1 will be 0 and the position of the first row with value 0 will be 102.
